### ISSUE TITLE
autocompletion fails after re-sourcing vimrc

### ISSUE THREAD

**Issue by reesew** [garbas/vim-snipmate#220]
I'm not entirely sure this is a snipmate issue, and I do a bunch of research every time I run into it, but I've found nothing so far. I'm currently on commit 7f91de3

Steps: 
1. type `if<tab>`, autocompletion works
2. run `:so ~/.vimrc`
3. type `if<tab>`, autocompletion fails with errors
```
Error detected while processing function snipMate#TriggerSnippet..snipMate#GetSnippetsForWordBelowCursor:
line   31:
E716: Key not present in Dictionary: get_snippets
Press ENTER or type command to continue
Error detected while processing function snipMate#TriggerSnippet..snipMate#GetSnippetsForWordBelowCursor:
line   31:
E116: Invalid arguments for function funcref#Call(g:snipMate['get_snippets'], [snipMate#ScopesByFile(), word]))
Press ENTER or type command to continue
Error detected while processing function snipMate#TriggerSnippet..snipMate#GetSnippetsForWordBelowCursor:
line   31:
E116: Invalid arguments for function items(funcref#Call(g:snipMate['get_snippets'], [snipMate#ScopesByFile(), word]))
Press ENTER or type command to continue
Error detected while processing function snipMate#TriggerSnippet..snipMate#GetSnippetsForWordBelowCursor:
line   31:
E15: Invalid expression: items(funcref#Call(g:snipMate['get_snippets'], [snipMate#ScopesByFile(), word]))
```

My .vimrc is pretty long, so I won't include it all here, but it has the following lines for vundle
```
  Bundle "MarcWeber/vim-addon-mw-utils"
  Bundle "tomtom/tlib_vim"
  Bundle "garbas/vim-snipmate"
  Bundle "honza/vim-snippets"
```
It also includes
```
let g:snipMate = {}
let g:snipMate.scope_aliases = {}
let g:snipMate.scope_aliases['ruby'] = 'ruby,rails'
```

Is there anything known about the order required for the vimrc lines related to loading snipmate? Or things that would break it in this way?
My workaround has always been to close and reopen vim

**Comment by reesew** [comment]
I can't believe I made progress on this after so many months or years. Removing the lines 
```
let g:snipMate = {}
let g:snipMate.scope_aliases = {}
let g:snipMate.scope_aliases['ruby'] = 'ruby,rails'
```
made a new issue show up (truncated, as it was hundreds of lines long)

```
Error detected while processing /home/reesew/.vim/bundle/vim-snippets/snippets/javascript/javascript-jquery.snippets:
line    1:
E492: Not an editor command: snippet add
Press ENTER or type command to continue
Error detected while processing /home/reesew/.vim/bundle/vim-snippets/snippets/javascript/javascript-jquery.snippets:
line    2:
E492: Not an editor command: ^I${1:obj}.add('${2:selector expression}')
Press ENTER or type command to continue
Error detected while processing /home/reesew/.vim/bundle/vim-snippets/snippets/javascript/javascript-jquery.snippets:
line    3:
E492: Not an editor command: snippet addClass
Press ENTER or type command to continue
Error detected while processing /home/reesew/.vim/bundle/vim-snippets/snippets/javascript/javascript-jquery.snippets:
line    4:
E492: Not an editor command: ^I${1:obj}.addClass('${2:class name}')
Press ENTER or type command to continue
Error detected while processing /home/reesew/.vim/bundle/vim-snippets/snippets/javascript/javascript-jquery.snippets:
line    5:
E492: Not an editor command: snippet after
```

Anyway, this issue seemed to be caused by the line `autocmd!` at the start of my .vimrc, which was fixed by starting the .vimrc with 
```
:augroup vimrc
:autocmd!
```
and ending with 
```
:augroup END
```

> Relevant portion of the manual:
> ```
> When your .vimrc file is sourced twice, the autocommands will appear twice.
> To avoid this, put this command in your .vimrc file, before defining
> autocommands:
> 
> 	:autocmd!	" Remove ALL autocommands for the current group.
> 
> If you don't want to remove all autocommands, you can instead use a variable
> to ensure that Vim includes the autocommands only once:
> 
> 	:if !exists("autocommands_loaded")
> 	:  let autocommands_loaded = 1
> 	:  au ...
> 	:endif
> 
> When the [group] argument is not given, Vim uses the current group (as defined
> with ":augroup"); otherwise, Vim uses the group defined with [group].  Note
> that [group] must have been defined before.  You cannot define a new group
> with ":au group ..."; use ":augroup" for that.
> ```
> 

Is snipmate using autocommands in the default group?

Anyway, all of the above autocmd stuff may unrelated. If so, I could start a new issue. 

After fixing the autocmd group stuff, and removing the lines for the scope_aliases, the problem went away. I still can't add the scope_alias lines without reproducing the original error

**Comment by ajzafar** [comment]
@reesew Try changing the `let g:snipMate = {}` line to `let g:snipMate = get(g:, 'snipMate', {})`. This will do nothing if `g:snipMate` already exists and create it if it doesn't. When SnipMate is first sourced, it creates `g:snipMate` if it doesn't exist and then adds stuff to it. I think when you source your vimrc, the dict is created anew. Since SnipMate isn't sourced again, the extra bits are lost. That's what the original errors are trying to say.

**Comment by reesew** [comment]
This works. Should the readme reflect that?
What about the autocmd group problem, should I create an issue for that?

**Comment by ajzafar** [comment]
Sorry I thought I already pushed that change. The autocommands are grouped now.

**Comment by mig-hub** [comment]
I have the exact same problem although I use "Plug" and the problem started to show after I added this line because I wanted to be able to use version 1 of the parser:

```
let g:snipMate = { 'snippet_version' : 1 }
```

This is the only config I have for snipmate. It does use version 1 now, but if I source my vimrc then all subsequent tab expansions fail similarly and it prints a "-1" in my buffer.

**Comment by mig-hub** [comment]
@ajzafar Thank you for your solution !
